graphql_compiler/compiler/ir_lowering_sql/__init__.py

```diff
@@ -1,35 +1,148 @@
+                    u'parent_location.query_path, optional_scopes_depth, recursive_scopes_depth '
+                    u'and types to be equal for LocationInfos sharing the same query path.'.format(
+                        location.query_path, location_info, equivalent_location_info))
+
+        query_path_to_location_info[location.query_path] = location_info
+    return query_path_to_location_info
+
+
+def _location_infos_equal(left, right):
+    """Return True if LocationInfo objects are equivalent for the SQL backend, False otherwise."""
+    if not isinstance(left, LocationInfo) or not isinstance(right, LocationInfo):
+        raise AssertionError(
+            (u'Unsupported LocationInfo comparison between types {} and {} '
+             u'with values {}, {}').format(type(left), type(right), left, right))
+    optional_scopes_depth_equal = (left.optional_scopes_depth == right.optional_scopes_depth)
+
+    parent_query_paths_equal = (
+        (left.parent_location is None and right.parent_location is None) or
+        (left.parent_location.query_path == right.parent_location.query_path))
+
+    recursive_scopes_depths_equal = (left.recursive_scopes_depth == right.recursive_scopes_depth)
+
+    types_equal = left.type == right.type
+
+    all_equal = (optional_scopes_depth_equal and parent_query_paths_equal and
+                 recursive_scopes_depths_equal and types_equal)
+    return all_equal
+
+
+def _map_query_path_to_outputs(construct_result, query_path_to_location_info):
+    """Assign the output fields of a ConstructResult block to their respective query_path."""
+    query_path_to_output_fields = {}
+    for output_name, field in six.iteritems(construct_result.fields):
+        field_name = field.location.field
+        if field_name in UNSUPPORTED_META_FIELDS:
+            raise exceptions.GraphQLNotSupportedByBackendError(
+                u'"{}" is unsupported for output.'.format(UNSUPPORTED_META_FIELDS[field_name]))
+        output_query_path = field.location.query_path
+        output_field_info = SqlOutput(
+            field_name=field_name,
+            output_name=output_name,
+            graphql_type=query_path_to_location_info[output_query_path].type)
+        output_field_mapping = query_path_to_output_fields.setdefault(output_query_path, [])
+        output_field_mapping.append(output_field_info)
+    return query_path_to_output_fields
+
+
+def _map_block_index_to_location(ir_blocks):
+    """Associate each IR block with it's corresponding location, by index."""
+    block_index_to_location = {}
+    current_block_ixs = []
+    for num, ir_block in enumerate(ir_blocks):
+        current_block_ixs.append(num)
+        if isinstance(ir_block, blocks.MarkLocation):
+            for ix in current_block_ixs:
+                block_index_to_location[ix] = ir_block.location
+            current_block_ixs = []
+    return block_index_to_location
```

Comment:
Since `ir_blocks` does not end in a `MarkLocation` block, doesn't this mean that a number of blocks at the very end of the `ir_blocks` list will not be associated with a location?

Consider writing an assertion that `block_index_to_location` has exactly as many entries as `ir_blocks`, to be sure.

Reply:
This is a good point. The assertion I arrived at was that any blocks after global operations start do not require location info for the SQL backend. To get there, I skip everything after global operations and make sure there was nothing left to be associated with a location before global operations. Does that seem reasonable?